src/audio/asrc/asrc_common.c

```diff
@@ -0,0 +1,799 @@
+	n = cd->source_frames * audio_stream_get_channels(source);
+	while (n > 0) {
+		n_wrap_src = (int16_t *)audio_stream_get_end_addr(source) - src;
+		n_copy = (n < n_wrap_src) ? n : n_wrap_src;
+		s_copy = n_copy * sizeof(int16_t);
+		ret = memcpy_s(buf, s_copy, src, s_copy);
+		assert(!ret);
+
+		/* Update and check both source and destination for wrap */
+		n -= n_copy;
+		src += n_copy;
+		buf += n_copy;
+		src_inc_wrap_s16(&src, audio_stream_get_end_addr(source),
+				 audio_stream_get_size(source));
+	}
+
+	/* Run ASRC */
+	in_frames = cd->source_frames;
+	out_frames = cd->sink_frames;
+
+	if (cd->mode == ASRC_OM_PUSH)
+		ret = asrc_process_push16(dev, cd->asrc_obj,
+					  (int16_t **)cd->ibuf, &in_frames,
+					  (int16_t **)cd->obuf, &out_frames,
+					  &idx, 0);
+	else
+		ret = asrc_process_pull16(dev, cd->asrc_obj,
+					  (int16_t **)cd->ibuf, &in_frames,
+					  (int16_t **)cd->obuf, &out_frames,
+					  in_frames, &idx);
+
+	if (ret)
+		comp_err(dev, "src_copy_s16(), error %d", ret);
+
+	buf = (int16_t *)cd->obuf[0];
+	n = out_frames * audio_stream_get_channels(sink);
+	while (n > 0) {
+		n_wrap_snk = (int16_t *)audio_stream_get_end_addr(sink) - snk;
+		n_copy = (n < n_wrap_snk) ? n : n_wrap_snk;
+		s_copy = n_copy * sizeof(int16_t);
+		ret = memcpy_s(snk, s_copy, buf, s_copy);
+		assert(!ret);
+
+		/* Update and check both source and destination for wrap */
+		n -= n_copy;
+		snk += n_copy;
+		buf += n_copy;
+		src_inc_wrap_s16(&snk, audio_stream_get_end_addr(sink),
+				 audio_stream_get_size(sink));
+	}
+
+	*n_read = in_frames;
+	*n_written = out_frames;
+}
+
+static int asrc_initialize_buffers(struct asrc_farrow *src_obj)
+{
+	int32_t *buf_32;
+	int16_t *buf_16;
+	int ch;
+	size_t buffer_size;
+
+	/* Set buffer_length to filter_length * 2 to compensate for
+	 * missing element wise wrap around while loading but allowing
+	 * aligned loads.
+	 */
+	src_obj->buffer_length = src_obj->filter_length * 2;
+	src_obj->buffer_write_position = src_obj->filter_length;
+
+	if (src_obj->bit_depth == 32) {
+		buffer_size = src_obj->buffer_length * sizeof(int32_t);
+
+		for (ch = 0; ch < src_obj->num_channels; ch++) {
+			buf_32 = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, buffer_size);
+
+			if (!buf_32)
+				return -ENOMEM;
+
+			src_obj->ring_buffers32[ch] = buf_32;
+		}
+	} else {
+		buffer_size = src_obj->buffer_length * sizeof(int16_t);
+
+		for (ch = 0; ch < src_obj->num_channels; ch++) {
+			buf_16 = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, buffer_size);
+
+			if (!buf_16)
+				return -ENOMEM;
+
+			src_obj->ring_buffers16[ch] = buf_16;
+		}
+	}
+
+	return 0;
+}
+
+static void asrc_release_buffers(struct asrc_farrow *src_obj)
+{
+	int32_t *buf_32;
+	int16_t *buf_16;
+	int ch;
+
+	if (!src_obj)
+		return;
+
+	if (src_obj->bit_depth == 32)
+		for (ch = 0; ch < src_obj->num_channels; ch++) {
+			buf_32 = src_obj->ring_buffers32[ch];
+
+			if (buf_32) {
+				rfree(buf_32);
+				src_obj->ring_buffers32[ch] = NULL;
```

Comment:
assuming no race is possible, this could be as simple as
```
for (ch = 0; ch < src_obj->num_channels; ch++) {
	rfree(src_obj->ring_buffers32[ch]);
	src_obj->ring_buffers32[ch] = NULL;
}
```
but if this is indeed trying to protect against a race, then it should first set address to `NULL` and then free memory. Otherwise unclear why these temporary variables are needed. Same below